### PR TITLE
Data race in tests: TestApplicationsUpgradeOverREST race

### LINKED ISSUE #2507: TestApplicationsUpgradeOverREST race
## Summary
```
WARNING: DATA RACE
Read at 0x00c0002d46a1 by goroutine 45:
  testing.tRunner.func1()
      /usr/local/go/src/testing/testing.go:1100 +0xe8
  runtime.call16()
      /usr/local/go/src/runtime/asm_amd64.s:550 +0x3d
  testing.(*T).FailNow()
      <autogenerated>:1 +0x44
  github.com/algorand/go-algorand/test/framework/fixtures.(*synchTest).FailNow()
      /Users/bolson/gopath/src/github.com/algorand/go-algorand/test/framework/fixtures/fixture.go:115 +0x9b
  github.com/stretchr/testify/require.NoError()
      /Users/bolson/gopath/pkg/mod/github.com/stretchr/testify@v1.6.1/require/require.go:1038 +0x104
  github.com/stretchr/testify/require.(*Assertions).NoError()
      /Users/bolson/gopath/pkg/mod/github.com/stretchr/testify@v1.6.1/require/require_forward.go:813 +0xd0
  github.com/algorand/go-algorand/test/e2e-go/upgrades.TestApplicationsUpgradeOverREST()
      /Users/bolson/gopath/src/github.com/algorand/go-algorand/test/e2e-go/upgrades/application_support_test.go:153 +0xe06
  testing.tRunner()
      /usr/local/go/src/testing/testing.go:1194 +0x202

Previous write at 0x00c0002d46a1 by goroutine 91:
  testing.(*common).FailNow()
      /usr/local/go/src/testing/testing.go:741 +0x4f
  testing.(*T).FailNow()
      <autogenerated>:1 +0x44
  github.com/algorand/go-algorand/test/framework/fixtures.(*synchTest).FailNow()
      /Users/bolson/gopath/src/github.com/algorand/go-algorand/test/framework/fixtures/fixture.go:115 +0x9b
  github.com/stretchr/testify/require.NoError()
      /Users/bolson/gopath/pkg/mod/github.com/stretchr/testify@v1.6.1/require/require.go:1038 +0x104
  github.com/algorand/go-algorand/test/framework/fixtures.(*LibGoalFixture).nodeExitWithError()
      /Users/bolson/gopath/src/github.com/algorand/go-algorand/test/framework/fixtures/libgoalFixture.go:125 +0x305
  github.com/algorand/go-algorand/test/framework/fixtures.(*LibGoalFixture).nodeExitWithError-fm()
      /Users/bolson/gopath/src/github.com/algorand/go-algorand/test/framework/fixtures/libgoalFixture.go:107 +0x68
  github.com/algorand/go-algorand/nodecontrol.(*NodeController).StartAlgod.func2()
      /Users/bolson/gopath/src/github.com/algorand/go-algorand/nodecontrol/algodControl.go:232 +0xe8

Goroutine 45 (running) created at:
  testing.(*T).Run()
      /usr/local/go/src/testing/testing.go:1239 +0x5d7
  testing.runTests.func1()
      /usr/local/go/src/testing/testing.go:1512 +0xa6
  testing.tRunner()
      /usr/local/go/src/testing/testing.go:1194 +0x202
  testing.runTests()
      /usr/local/go/src/testing/testing.go:1510 +0x612
  testing.(*M).Run()
      /usr/local/go/src/testing/testing.go:1418 +0x3b3
  main.main()
      _testmain.go:57 +0x236

Goroutine 91 (finished) created at:
  github.com/algorand/go-algorand/nodecontrol.(*NodeController).StartAlgod()
      /Users/bolson/gopath/src/github.com/algorand/go-algorand/nodecontrol/algodControl.go:224 +0x627
  github.com/algorand/go-algorand/netdeploy.Network.startNodes()
      /Users/bolson/gopath/src/github.com/algorand/go-algorand/netdeploy/network.go:339 +0x357
  github.com/algorand/go-algorand/netdeploy.Network.Start()
      /Users/bolson/gopath/src/github.com/algorand/go-algorand/netdeploy/network.go:293 +0x504
  github.com/algorand/go-algorand/test/framework/fixtures.(*LibGoalFixture).Start()
      /Users/bolson/gopath/src/github.com/algorand/go-algorand/test/framework/fixtures/libgoalFixture.go:264 +0xfa
  github.com/algorand/go-algorand/test/framework/fixtures.(*LibGoalFixture).setup()
      /Users/bolson/gopath/src/github.com/algorand/go-algorand/test/framework/fixtures/libgoalFixture.go:101 +0x551
  github.com/algorand/go-algorand/test/framework/fixtures.(*LibGoalFixture).Setup()
      /Users/bolson/gopath/src/github.com/algorand/go-algorand/test/framework/fixtures/libgoalFixture.go:67 +0x91
  github.com/algorand/go-algorand/test/framework/fixtures.(*RestClientFixture).Setup()
      /Users/bolson/gopath/src/github.com/algorand/go-algorand/test/framework/fixtures/restClientFixture.go:43 +0x8b
  github.com/algorand/go-algorand/test/e2e-go/upgrades.TestApplicationsUpgradeOverREST()
      /Users/bolson/gopath/src/github.com/algorand/go-algorand/test/e2e-go/upgrades/application_support_test.go:78 +0x1c5
  testing.tRunner()
      /usr/local/go/src/testing/testing.go:1194 +0x202
```

## Scope/Requirements
Make test pass `go test -race` reliably.

## Urgency/Relative Priority
This may be a flakey test slowing down validations and development.

### PR DESCRIPTION
**Pull request by algonautshant**
## Summary

When multiple threads use the synchTest object, and one fails, all call FailNow.

After the first call, the main test routine reads t.finished to terminate the test.

While the main test routine is reading t.finished, another thread calls FailNow, which in turn writes to t.finished, causing the data race.

While FailNow called from synchTest are guarded by a mutex, the main testing object has the TestingTB object, which does not use the same mutex.

The solution is to call FailNow from synchTest object only the first time.

fixes #2507

### More details:
The problem with the string of data race conditions we are observing in different tests when the test fails is the following:

When the test fails, FailNow in go/src/testing/testing.go sets `c.finished`, which is the "Previous write" part of the data race
```
func (c *common) FailNow() {
...
	c.finished = true
...
```
this call comes from different threads, possibly multiple ones, forked from (not exclusively) 
go-algorand/nodecontrol/algodControl.go
```
func (nc *NodeController) StartAlgod(args AlgodStartArgs) (alreadyRunning bool, err error) {
...
	go func() {
...
			if args.ExitErrorCallback != nil {
				args.ExitErrorCallback(nc, err)
```
where, `ExitErrorCallback` is:
```
func (f *LibGoalFixture) nodeExitWithError(nc *nodecontrol.NodeController, err error) {
...
		require.NoError(f.t, err, "Node at %s has terminated with an error", nc.GetDataDir())
...
}
```
and NoError, in case of an error, calls `FailNow`.

`f.t` happens to be `synchTest`, which has a mutex guarding the `FailNow` call, and hence, prevents simultaneous access `t.finished`.

However, there is one more read access to `t.finished`, which comes from the following location, which does not use synchTest object, hence, does not have the mutex used for the write:

go/src/testing/testing.go
```
func tRunner(t *T, fn func(t *T)) {
...
	defer func() {
...
		if !t.finished && err == nil {
```

In other words, when there are multiple threads, and all are fine, no thread will write to `t.finished`. When something goes wrong and `FailNow` is initiated, the other threads will be notified of the failure, and, in their turn, also report `require.NoError`, which calls `FailNow`, and writes `t.finished = true`.

### CODE REVIEW

**Comment by tsachiherman** [comment]
Based on the details you've provided, I don't believe that the proposed solution would fix the issue:
Consider the case where a new AlgodController is being created, and it's getting returned with an error *after* the test ends (without reporting any issue ). It would still call the "t.Fail", which would datarace with the tRunner.
I think that we should disable all Fail calls once we've completed the test execution. We can do that using a destructor, in a similar way to `SetTestContext()`

**Comment by algonautshant** [comment]
You are right. My changes addressed only failures caused by the test, but not failures coming from the fixture. 
This change should address it. I still need to update other tests with this.

File: test/framework/fixtures/fixture.go [hunk 1]
```diff
@@ -86,7 +86,18 @@ func SynchronizedTest(tb TestingTB) TestingTB {
 
 type synchTest struct {
 	deadlock.Mutex
-	t TestingTB
+	t                  TestingTB
+	dontReportFailures bool
+}
+
+// TerminateTestFailures should be called within each test using a shared fixture.
```
Comment (tsachiherman) [inline test/framework/fixtures/fixture.go:93]: I'm not sure if that's true only for shared fixtures.

File: test/framework/fixtures/fixture.go [hunk 2]
```diff
@@ -107,12 +118,18 @@ func (st *synchTest) Errorf(format string, args ...interface{}) {
...
+	if !st.dontReportFailures {
+		st.dontReportFailures = true
+		st.t.Fail()
+	}
 }
 func (st *synchTest) FailNow() {
 	st.Lock()
 	defer st.Unlock()
-	st.t.FailNow()
+	if !st.dontReportFailures {
```
Comment (tsachiherman) [inline test/framework/fixtures/fixture.go:129]: we'll need the same for Error/Errorf/Fatal/Fatalf/Skip***

File: test/framework/fixtures/fixture.go [hunk 3]
```diff
@@ -86,7 +86,18 @@ func SynchronizedTest(tb TestingTB) TestingTB {
...
 	deadlock.Mutex
-	t TestingTB
+	t                  TestingTB
+	dontReportFailures bool
+}
+
+// TerminateTestFailures should be called within each test using a shared fixture.
+// It ensures the base test will no longer get t.finished modified and cause a data race.
+// It should be called in the form of "defer fixtures.TerminateTestFailures(t)"
+func TerminateTestFailures(t TestingTB) {
```
Comment (tsachiherman) [inline test/framework/fixtures/fixture.go:96]: maybe `ShutdownSynchronizedTest` ?

**Review by tsachiherman — CHANGES_REQUESTED** [review]
looks good. make sure to fix
```sh
Go files must be formatted with gofmt. Please run:
  gofmt -w /home/circleci/project/test/e2e-go/upgrades/application_support_test.go
```

**Review by tsachiherman — APPROVED** [review]
loos good.
